README.md
Pulse
=====

[![hex.pm version](https://img.shields.io/hexpm/v/pulse.svg?style=flat)](https://hex.pm/packages/pulse)

Service registration and discovery library for Elixir. Relies on [etcd](https://coreos.com/etcd/) as an external service registry.

Works best with FQDN node names, such as `my_node@ip-123-234-124-235.local`.

### Installation

The latest version is `0.1.0` and requires Elixir `~> 1.2`. New releases may change this minimum compatible version depending on breaking language changes. The [changelog](https://github.com/heroiclabs/pulse/blob/master/CHANGELOG.md) lists every available release and its corresponding language version requirement.

Releases are published through [hex.pm](https://hex.pm/packages/pulse). Add as a dependency in your `mix.exs` file:
```elixir
defp deps do
  [ { :pulse, "~> 0.1" } ]
end
```

Also ensure it's listed in the `mix.exs` list of applications to start:
```elixir
def application do
  [
    applications: [:pulse]
  ]
end
```

### Configuration

Below is the complete default configuration. All parameters can be changed.

```elixir
config :pulse,
  directory: "pulse"
```

For `etcd` connection configuration, see [Sonic](https://github.com/heroiclabs/sonic).

### Usage

There are three discrete functions handled by Pulse:

* Service Registration - Publishing service status so it is discoverable by other nodes.
* Service Discovery - Retrieving a list of available nodes registered to provide a service.
* Service Directory - Maintaining the internal service registry and handling connections to discovered nodes.

For the examples below, we'll assume a typical `Application` module with a start function:

```elixir
def start(_type, _args) do
import Supervisor.Spec, warn: false

children = [
  worker(MyApp.SomeWorker, [])
]

opts = [strategy: :one_for_one, name: MyApp.Supervisor]
Supervisor.start_link(children, opts)
end
```

#### Service Registration

`Pulse.Register` is responsible for service registration. To use it, declare it as a worker in your application's supervision tree.

The module takes 4 parameters in a keyword list, for legibility:

* `service` - A binary indicating the service name to register.
* `ttl` - An integer indicating the number of seconds the registration should live unless refreshed.
* `heartbeat` - An integer indicating the number of seconds to wait between registration refreshes.
* `delay` - An integer indicating the number of seconds to wait before first registration.

```elixir
children = [
  worker(Pulse.Register, [[service: "my_service", ttl: 15, heartbeat: 5, delay: 5]])
]
```

Each `Pulse.Register` worker is only responsible for one service, but you can register the application as providing multiple services by declaring more than one worker.

```elixir
children = [
  worker(Pulse.Register, [[service: "my_service", ttl: 15, heartbeat: 5, delay: 5]], id: MyApp.MyServiceRegister),
  worker(Pulse.Register, [[service: "my_other_service", ttl: 15, heartbeat: 5, delay: 5]], id: MyApp.MyOthereSrviceRegister)
  # ...
]
```

The configuration options should be tuned for your application, however the values shown above are a good baseline.

#### Service Discovery

Start one or more `Pulse.Discover` workers to look up registered nodes for required services.

The module taks 3 parameters in a keyword list, for legibility:

* `service` - A binary indicating the service name to discover.
* `poll` - An integer indicating the number of seconds to wait between discovery refreshes.
* `delay` - An integer indicating the number of seconds to wait before first discovery.

```elixir
children = [
  worker(Pulse.Discover, [[service: "my_service", poll: 5, delay: 1]])
]
```

Each `Pulse.Discover` worker is only responsible for one service, but you can discover multiple services by declaring more than one worker.

```elixir
children = [
  worker(Pulse.Discover, [[service: "my_service", poll: 5, delay: 1]], id: MyApp.MyServiceDiscover),
  worker(Pulse.Discover, [[service: "my_other_service", poll: 5, delay: 1]], id: MyApp.MyOtherServiceDiscover)
  # ...
]
```

The configuration options should be tuned for your application, however the values shown above are a good baseline.

#### Service Directory

The `Pulse.Directory` module maintains connections and tracks available nodes registered for each discovered service. This module is monitored by Pulse internally, you do not need to start a worker to use it.

`Pulse.Directory.get/1` is the primary way to retrieve available nodes for any given discovered service.

```elixir
iex> Pulse.Directory.get("my_service")
[:"my_machine1@ip-123-234-124-235.local", :"my_machine2@ip-123-234-124-235.local"]
```

These nodes will already be connected by the `Pulse.Directory` process and are valid RPC targets if necessary. The directory also monitors connections and will unregister nodes from service lists if they disconnect.

### License

```none
Copyright 2016 Heroic Labs

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
Changelog
=========

### Unreleased changes

None.

---

### v0.1.0, 18 Mar 2016, Elixir `~> 1.2`

* Initial release.


mix.exs
defmodule Pulse.Mixfile do
  use Mix.Project

  @version "0.1.3"

  def project do
    [app: :pulse,
     version: @version,
     elixir: "~> 1.2",
     deps: deps,
     package: package,

     name: "Pulse",
     docs: [extras: ["README.md", "CHANGELOG.md"],
            main: "readme",
            source_ref: "v#{@version}"],
     source_url: "https://github.com/heroiclabs/pulse",
     description: description]
  end

  # Application configuration.
  def application do
    [
      mod: {Pulse, []},
      applications: [:logger, :sonic],
      env: [
        directory: "pulse"
      ]
    ]
  end

  # List of dependencies.
  defp deps do
    [{:sonic, "~> 0.1"},

     # Docs
     {:ex_doc, "~> 0.11", only: :dev},
     {:earmark, "~> 0.2", only: :dev}]
  end

  # Description.
  defp description do
    """
    Service registration and discovery library for Elixir. Relies on etcd as an external service registry.
    """
  end

  # Package info.
  defp package do
    [files: ["lib", "mix.exs", "README.md", "LICENSE"],
     maintainers: ["Andrei Mihu"],
     licenses: ["Apache 2.0"],
     links: %{"GitHub" => "https://github.com/heroiclabs/pulse"}]
  end

end


lib/pulse/discover.ex
defmodule Pulse.Discover do
  use GenServer
  require Logger

  @moduledoc """
  Start one or more `Pulse.Discover` workers to look up registered nodes for required services.

  The module taks 3 parameters in a keyword list, for legibility:

  * `service` - A binary indicating the service name to discover.
  * `poll` - An integer indicating the number of seconds to wait between discovery refreshes.
  * `delay` - An integer indicating the number of seconds to wait before first discovery.

  ```elixir
  children = [
    worker(Pulse.Discover, [[service: "my_service", poll: 5, delay: 1]])
  ]
  ```

  Each `Pulse.Discover` worker is only responsible for one service, but you can discover multiple services by declaring more than one worker.

  ```elixir
  children = [
    worker(Pulse.Discover, [[service: "my_service", poll: 5, delay: 1]], id: MyApp.MyServiceDiscover),
    worker(Pulse.Discover, [[service: "my_other_service", poll: 5, delay: 1]], id: MyApp.MyOtherServiceDiscover)
    # ...
  ]
  ```

  The configuration options should be tuned for your application, however the values shown above are a good baseline.
  """

  @doc false
  def start_link(args) do
    GenServer.start_link(__MODULE__, args, [])
  end

  def init([service: service, poll: poll, delay: delay]) do
    pid = self()

    # Schedule the first discover refresh.
    Process.send_after(pid, :discover, delay * 1000)

    {:ok, [pid: pid, service: service, poll: poll]}
  end

  def handle_info(:discover, [pid: pid, service: service, poll: poll] = state) do
    # Determine what the discover path will be.
    path = [Application.get_env(:pulse, :directory), service]

    # Execute the discover request and handle the result.
    result = Sonic.Client.dir_list(path)
    case result do
      {:ok, 200, _headers, body} ->
        nodes = Enum.map(body["node"]["nodes"] || [], fn(node) ->
            node["value"] |> String.to_atom
          end)
        Pulse.Directory.update(service, nodes)
      {:ok, 404, _headers, _body} ->
        :ok
      _ ->
        Logger.error("Pulse.Discover failed for service #{service}: #{inspect result}")
    end

    # Schedule the next discover refresh.
    Process.send_after(pid, :discover, poll * 1000)
    {:noreply, state}
  end

end


lib/pulse/register.ex
defmodule Pulse.Register do
  use GenServer
  require Logger

  @moduledoc """
  `Pulse.Register` is responsible for service registration. To use it, declare it as a worker in your application's supervision tree.

  The module takes 4 parameters in a keyword list, for legibility:

  * `service` - A binary indicating the service name to register.
  * `ttl` - An integer indicating the number of seconds the registration should live unless refreshed.
  * `heartbeat` - An integer indicating the number of seconds to wait between registration refreshes.
  * `delay` - An integer indicating the number of seconds to wait before first registration.

  ```elixir
  children = [
    worker(Pulse.Register, [[service: "my_service", ttl: 15, heartbeat: 5, delay: 5]])
  ]
  ```

  Each `Pulse.Register` worker is only responsible for one service, but you can register the application as providing multiple services by declaring more than one worker.

  ```elixir
  children = [
    worker(Pulse.Register, [[service: "my_service", ttl: 15, heartbeat: 5, delay: 5]], id: MyApp.MyServiceRegister),
    worker(Pulse.Register, [[service: "my_other_service", ttl: 15, heartbeat: 5, delay: 5]], id: MyApp.MyOthereSrviceRegister)
    # ...
  ]
  ```

  The configuration options should be tuned for your application, however the values shown above are a good baseline.
  """

  @doc false
  def start_link(args) do
    GenServer.start_link(__MODULE__, args, [])
  end

  def init([service: service, ttl: ttl, heartbeat: heartbeat, delay: delay]) do
    pid = self()

    # Schedule the first registration.
    Process.send_after(pid, :register, delay * 1000)

    {:ok, [pid: pid, service: service, ttl: ttl, heartbeat: heartbeat]}
  end

  def handle_info(:register, [pid: pid, service: service, ttl: ttl, heartbeat: heartbeat] = state) do
    # Determine what the registration path and content will be.
    id = Node.self |> to_string
    path = [Application.get_env(:pulse, :directory), service, id]

    # Execute the registration and report the result.
    result = Sonic.Client.kv_put(path, id, ttl: ttl)
    case result do
      {:ok, status, _headers, _body} when status == 200 or status == 201 ->
        :ok
      _ ->
        Logger.error("Pulse.Register failed for service #{service}: #{inspect result}")
    end

    # Schedule the next registration heartbeat.
    Process.send_after(pid, :register, heartbeat * 1000)
    {:noreply, state}
  end

end


lib/pulse/directory.ex
defmodule Pulse.Directory do
  use GenServer
  require Logger

  @moduledoc """
  The `Pulse.Directory` module maintains connections and tracks available nodes registered for each discovered service. This module is monitored by Pulse internally, you do not need to start a worker to use it.

  `Pulse.Directory.get/1` is the primary way to retrieve available nodes for any given discovered service.

  ```elixir
  > Pulse.Directory.get("my_service")
  [:"my_machine1@ip-123-234-124-235.local", :"my_machine2@ip-123-234-124-235.local"]
  ```

  These nodes will already be connected by the `Pulse.Directory` process and are valid RPC targets if necessary. The directory also monitors connections and will unregister nodes from service lists if they disconnect.
  """

  @doc false
  def update(service, nodes_snapshot) do
    GenServer.call(__MODULE__, {:update, service, nodes_snapshot})
  end

  @doc """
  Retrieve a list of connected nodes that are registered to provide the given service.

  ```elixir
  > Pulse.Directory.get("my_service")
  [:"my_machine1@ip-123-234-124-235.local", :"my_machine2@ip-123-234-124-235.local"]
  ```
  """
  def get(service) do
    GenServer.call(__MODULE__, {:get, service})
  end

  #
  # Internal.
  #

  @doc false
  def start_link() do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
  end

  def init([]) do
    {:ok, %{"services" => %{}, "nodes" => %{}}}
  end

  def handle_info({:nodedown, node}, %{"services" => services, "nodes" => nodes}) do
    Logger.warn("Pulse.Directory unexpected disconnect from node #{node}")

    # Update the services-to-nodes and nodes-to-services mappings.
    services = Enum.into(services, %{}, fn({service, service_nodes}) ->
        {service, List.delete(service_nodes, node)}
      end)
    nodes = Map.delete(nodes, node)

    {:noreply, %{"services" => services, "nodes" => nodes}}
  end

  def handle_call({:get, service}, _from, %{"services" => services} = state) do
    {:reply, services[service] || [], state}
  end
  def handle_call({:update, service, nodes_snapshot}, _from, %{"services" => services, "nodes" => nodes}) do
    service_nodes = services[service] || []

    # Connect new nodes as needed and update the state.
    new_nodes = complement(nodes_snapshot, service_nodes)
    {service_nodes, nodes} = Enum.reduce(new_nodes, {service_nodes, nodes}, fn(new_node, {service_nodes, nodes}) ->
        node_services = nodes[new_node]
        case node_services do
          # There are known services for this node, so assume it's a new node.
          nil ->
            case Node.connect(new_node) do
              true ->
                case Node.monitor(new_node, true) do
                  true ->
                    Logger.info("Pulse.Directory connected to node #{new_node}")
                    Logger.info("Pulse.Directory node #{new_node} registered for service #{service}")
                    {[new_node | service_nodes], Map.put(nodes, new_node, [service])}
                  _ ->
                    # Don't retry connection errors, the next directory update will reconnect.
                    Logger.warn("Pulse.Directory failed to monitor node #{new_node}")
                    {service_nodes, nodes}
                end
              _ ->
                # Don't retry connection errors, the next directory update will reconnect.
                Logger.warn("Pulse.Directory failed to connect to node #{new_node}")
                {service_nodes, nodes}
            end
          # This node is already connected, possibly by another service.
          _ ->
            Logger.info("Pulse.Directory node #{new_node} registered for service #{service}")
            {[new_node | service_nodes], Map.put(nodes, new_node, [service | node_services])}
        end
      end)

    # Disconnect missing nodes as needed.
    missing_nodes = complement(service_nodes, nodes_snapshot)
    {service_nodes, nodes} = Enum.reduce(missing_nodes, {service_nodes, nodes}, fn(missing_node, {service_nodes, nodes}) ->
        node_services = nodes[missing_node]
        case node_services do
          # Theis node is currently only serving this one service, disconnect.
          [^service] ->
            Node.monitor(missing_node, false)
            Node.disconnect(missing_node)
            Logger.info("Pulse.Directory node #{missing_node} unregistered for service #{service}")
            Logger.info("Pulse.Directory disconnected node #{missing_node}")
            {List.delete(service_nodes, missing_node), Map.delete(nodes, missing_node)}
          # There are other services for this node, do not disconnect.
          _ ->
            Logger.info("Pulse.Directory node #{missing_node} unregistered for service #{service}")
            {List.delete(service_nodes, missing_node), Map.put(nodes, missing_node, List.delete(node_services, service))}
        end
      end)

    # Update the services-to-nodes mapping.
    # The nodes-to-services mapping was updated through the new_nodes and missing_nodes processing above.
    services = case service_nodes do
      [] ->
        Map.delete(services, service)
      _ ->
        Map.put(services, service, service_nodes)
    end

    {:reply, :ok, %{"services" => services, "nodes" => nodes}}
  end

  #
  # Utilities.
  #

  defp complement(list1, list2) do
    List.foldl(list2, list1, fn(item, acc) ->
        List.delete(acc, item)
      end)
  end

end
